---
title: "TrioSim Vegnette"
author: "Min Shi"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

*TrioSim* is a  package that can simulate genotypes for case-parent triads with realistic linkage diequilibrium structure and allele frequency distribution by resampling triad genotypes from existing data.

## Main function ` TrioSim` 

` TrioSim` is the main function to simulate realistic trio genotypes. The example function call below simulates genotype data for 1000 case-parent triads under a genetic main effect scenario with a baseline disease prevalence of P0=0.001 and genetic relative risks of 1.5 and 2 for carrying the first and the second pathway respectively. This function call will  write PLINK output files into the current directory. Each set (.bim, .bed and .fam files) of PLINK files contain genotype data for one chromosome for all simulated trios. See documentation for more details.

 
```{r}
library(TrioSim)
 m.file <- paste(system.file(package = "TrioSim"),'/extdata/pop1_4chr_mom',sep='')
 f.file <- paste(system.file(package = "TrioSim"),'/extdata/pop1_4chr_dad',sep='')
 k.file <- paste(system.file(package = "TrioSim"),'/extdata/pop1_4chr_kid',sep='')
 input.plink.file <- c(m.file, f.file, k.file)

 TrioSim(input.plink.file, 'triad', fr.desire=0.05,pathways=list(1:4,5:8),n.ped=1000, N.brk=3, target.snp=NA,P0=0.001,is.OR=FALSE,risk.exposure= 1,risk.pathway.unexposed=c(1.5, 2), risk.pathway.exposed=c(1.5, 2), is.case=TRUE, e.fr=NA, pop1.frac=NA, rate.beta=0,rcmb.rate, no_cores=NA)

```


This following call simlates quantitative trait. The function will create 4 sets of plink files, one for each single chromosome.
```{r}
 TrioSim(input.plink.file, 'qtl', fr.desire=0.3,pathways=list(1:4,5:8),n.ped=1000, N.brk=3, target.snp=NA,P0=0.001,is.OR=FALSE,risk.exposure= 1,risk.pathway.unexposed=c(0.5, 1), risk.pathway.exposed=c(0.5, 1), is.case=TRUE, e.fr=NA, pop1.frac=NA, rate.beta=0,rcmb.rate, no_cores=5, qtl=T)
```

To simulate case-control data the function needs to be called twice, calls to simulate cases and controls respectively. The script below calls the function to simulate 1000 cases and 1000 controls. 
```{r}
## cases
 TrioSim(input.plink.file, 'case', fr.desire=0.05,pathways=list(1:4,5:8),n.ped=1000, N.brk=3, target.snp=NA,P0=0.001,is.OR=FALSE,risk.exposure= 1,risk.pathway.unexposed=c(1.5, 2), risk.pathway.exposed=c(1.5, 2), is.case=TRUE, e.fr=NA, pop1.frac=NA, rate.beta=0,rcmb.rate, no_cores=NA)
## controls
 TrioSim(input.plink.file, 'ctrl', fr.desire=0.05,pathways=list(1:4,5:8),n.ped=1000, N.brk=3, target.snp=NA,P0=0.001,is.OR=FALSE,risk.exposure= 1,risk.pathway.unexposed=c(1.5, 2), risk.pathway.exposed=c(1.5, 2), is.case=FALSE, e.fr=NA, pop1.frac=NA, rate.beta=0,rcmb.rate, no_cores=NA)

```



##Facility functions

The following set of functions are provided in case users want to have more controls over the simulation parameters. They are called by the main function to generate simulations. Users do not need to call them directly.

###`pick_target.snp`
Users can manually pick the target SNPs in the pathway or use the facility function `pick_target.snp` to pick the set of target SNPs in the pathway(s) based on a desired allele frequency. The example below uses the example files in the package to select 8 SNPs with allele frequencies close to 0.05. The function returns the selected target SNPs by giving the row numbers (i.e., the order) of the corresponding SNPs among all the SNPs in the associated "bim" file.

```{r}
 m.file <- paste(system.file(package = "TrioSim"),'/extdata/pop1_4chr_mom',sep='')
 f.file <- paste(system.file(package = "TrioSim"),'/extdata/pop1_4chr_dad',sep='')
 picked.target <- pick_target.snp(c(m.file,f.file),0.05, 8)
 cat('Target SNPs picked:',picked.target[[2]],'\n')
```


###`get.target.geno`
The function `get.target.geno` retrieves genotypes of the target SNPs and return the genotypes of the triads in a list of three elements: they are the observed genotypes of the mothers from family 1 to family n repeated twice, genotypes of the fathers from family 1 to family n repeated twice and genotypes of children from family 1 to n followed by (stacking on top of) genotypes of the complements at the target SNP.

```{r}
target.snp <- picked.target[[2]]
m.file <- paste(system.file(package = "TrioSim"),'/extdata/pop1_4chr_mom',sep='')
f.file <- paste(system.file(package = "TrioSim"),'/extdata/pop1_4chr_dad',sep='')
k.file <- paste(system.file(package = "TrioSim"),'/extdata/pop1_4chr_kid',sep='')
# the preloaded data frame snp.all2 contains the data frame read from the corresponding .bim file.
target.geno <- get.target.geno(c(m.file,f.file,k.file), target.snp,snp.all2)
```
The output *target.geno* is a list of three elements, each being a matrix of genotypes 
```{r}
length(target.geno)
```
For this example the genotypes is a 2000 x 8 numerical matrix
```{r}
mom.target <- target.geno[[1]]
dad.target <- target.geno[[2]]
kid.target <- target.geno[[3]]
str(mom.target)
```

To increase diversity, the program introduced breaking points at each chromosome. The break points can be picked manually or use function `get.brks`. The function tends to pick the breaking points at recombination hotspots if such data are passed in as a parameter *rcmb.rate*. 

```{r}
found.brks <- get.brks(N.brk=3,n.ped=1000, snp.all2, target.snp,rcmb.rate=rcmb.rate)
breaks <- found.brks[[1]]
family.position <- found.brks[[2]] 
```

This function returns a list of two items. The first is a 1000 x 17  matrix of integers showing where the chromosomal breaks is to take place for each individual in the simulated trios. Note that it also includes 0, the total number of SNPs and the breaking points between different chromosomes. Here 1000 denotes the number of triads in the simulated data as defined by the `n.ped` input parameter.

```{r}
dim(breaks)
head(breaks)
```

The second one is a 1000 x 8 matrix showing the chromosomal segments out of which each target SNP is selected for each simulated trio.

```{r}
dim(family.position)
head(family.position)
```

###`fit.risk.model.par`
`fit.risk.model.par` is a function that resamples families based on the specified risk model. It can simulate a homogenous scenario or a stratified scenario with two subpopulations. The risk model can involve exposure main effects as well as gene by exposure interations. This function is parallelized to shorten the running time. An example to call the function is given below.

```{r,fig.width=6, fig.height=6}
betas <- c(-6.4, 3.2, 5.8)
pwy <- list(1:4,5:8)
## scenarios of genetic main effects only for a binary phenotype
fitted.model1 <- fit.risk.model.par(n.ped=1000,brks=breaks,target.snp,fam.pos=family.position, 
mom.tar=mom.target,dad.tar=dad.target, kid.tar=kid.target, pathways=pwy, 
betas, e.fr=NA, betas,pop1.frac= NA,rate.beta=NA,qtl= FALSE)

sel.fam <- fitted.model1[[1]]
sim.pathway.geno <-  fitted.model1[[2]]

dim(sel.fam)
dim(sim.pathway.geno)

## scenarios of gene-environment interaction  for a binary phenotype
betas.e <- c(-6.4, 3.9, 6.5)

fitted.model2 <- fit.risk.model.par(n.ped=1000,brks=breaks,target.snp,fam.pos=family.position, 
mom.tar=mom.target,dad.tar=dad.target, kid.tar=kid.target, pathways=pwy, 
betas, e.fr= 0.2, betas.e,pop1.frac= NA,rate.beta=NA, qtl= FALSE)
exposure <-  fitted.model2[[3]]
table(exposure)

## scenarios of a quantitative trait phenotype
fitted.model3 <- fit.risk.model.par(n.ped=1000,brks=breaks,target.snp,fam.pos=family.position, 
mom.tar=mom.target,dad.tar=dad.target, kid.tar=kid.target, pathways=pwy, 
betas, e.fr=NA, betas,pop1.frac= NA,rate.beta=NA,qtl=TRUE)
qt.pheno <-  fitted.model3[[5]]

hist(qt.pheno,main='Histogram of Simulated Quantitative Trait',xlab='QT')
```

###`glue.chr.segment.par`
`glue.chr.segment.par` is a function that splices the triad chromosomal segments into "complete" trios. The spliced trio sets are written into separate plink
 files chromosome by chromosome. It is parallelized and if no no_cores value is given the total number
 of CPUs available will be used in the parallelization. Be aware that running the following code will write PLINK files to your current working direcotry.

```{r}
glue.chr.segment.par(c(m.file,f.file,k.file),'trio', breaks,sel.fam,snp.all2,sim.pathway.geno,target.snp,pop.vec=NA,no_cores=NA,flip=TRUE) 
```
