<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Min Shi" />

<meta name="date" content="2017-04-21" />

<title>TriadSim Vignette</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">TriadSim Vignette</h1>
<h4 class="author"><em>Min Shi</em></h4>
<h4 class="date"><em>2017-04-21</em></h4>



<p><em>TriadSim</em> is a package that can simulate genotypes for case-parent triads, case-control, and quantitative trait samples with realistic linkage diequilibrium structure and allele frequency distribution. For studies of epistasis one can simulate models that involve specific SNPs at specific sets of loci, which we will refer to as “pathways”. TriadSim generates genotype data by resampling triad genotypes from existing data. It takes genotypes in PLINK format as the input files. The genotypes for the mothers, fathers, and children are in separate files. The mothers, fathers, and children must be from the same set of triad families although the ordering of the families can be different for the three sets of data. After reading in the genotypes, a sorting step will reorder the families so that the three individuals in each family can realign.</p>
<div id="main-function-triadsim" class="section level2">
<h2>Main function <code>TriadSim</code></h2>
<p><code>TriadSim</code> is the main function to perform the simulations. The example function call below simulates genotype data for 1000 case-parent triads for 4 chromosomes (chromsomes 1, 8 17, 20) under a genetic main effect scenario with a baseline disease prevalence of P0=0.001 and genetic relative risks of 1.5 and 2 for carrying the first and the second pathway respectively. This function call will write output files in PLINK. The output file names and path to the directory are given by the parameter “out.put.file” and the chromosome number. Each set (.bim, .bed and .fam files) of PLINK files contain genotype data for one chromosome for all simulated samples. The name of the file is the concatenation of the value of the input parameter “out.put.file” and chromosome number. For example, if “out.put.file” is set to be “triad”, the names of the output files will be triad1, triad8, triad17 and triad20 for our example. See R package documentation for more details.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(TriadSim)
 m.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_mom'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 f.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_dad'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 k.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_kid'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 input.plink.file &lt;-<span class="st"> </span><span class="kw">c</span>(m.file, f.file, k.file)

 <span class="kw">TriadSim</span>(input.plink.file, <span class="dt">out.put.file=</span><span class="st">'triad'</span>, <span class="dt">fr.desire=</span><span class="fl">0.05</span>,<span class="dt">pathways=</span><span class="kw">list</span>(<span class="dv">1</span>:<span class="dv">4</span>,<span class="dv">5</span>:<span class="dv">8</span>),<span class="dt">n.ped=</span><span class="dv">1000</span>, <span class="dt">N.brk=</span><span class="dv">3</span>, <span class="dt">target.snp=</span><span class="ot">NA</span>,<span class="dt">P0=</span><span class="fl">0.001</span>,<span class="dt">is.OR=</span><span class="ot">FALSE</span>,<span class="dt">risk.exposure=</span> <span class="dv">1</span>,<span class="dt">risk.pathway.unexposed=</span><span class="kw">c</span>(<span class="fl">1.5</span>, <span class="dv">2</span>), <span class="dt">risk.pathway.exposed=</span><span class="kw">c</span>(<span class="fl">1.5</span>, <span class="dv">2</span>), <span class="dt">is.case=</span><span class="ot">TRUE</span>, <span class="dt">e.fr=</span><span class="ot">NA</span>, <span class="dt">pop1.frac=</span><span class="ot">NA</span>, <span class="dt">P0.ratio=</span><span class="dv">1</span>,rcmb.rate, <span class="dt">no_cores=</span><span class="ot">NA</span>)</code></pre></div>
<pre><code>## [1]  21 118 121 140 155 168 218 383
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to triad1.fam 
## Writing extended MAP file to triad1.bim 
## Writing BED file to triad1.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to triad8.fam 
## Writing extended MAP file to triad8.bim 
## Writing BED file to triad8.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to triad17.fam 
## Writing extended MAP file to triad17.bim 
## Writing BED file to triad17.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to triad20.fam 
## Writing extended MAP file to triad20.bim 
## Writing BED file to triad20.bed (SNP-major mode)</code></pre>
<p>The following call simulates a quantitative trait (by setting “qtl=T”). The function will create 4 sets of plink files, one for each chromosome.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">TriadSim</span>(input.plink.file, <span class="st">'qtl'</span>, <span class="dt">fr.desire=</span><span class="fl">0.3</span>,<span class="dt">pathways=</span><span class="kw">list</span>(<span class="dv">1</span>:<span class="dv">4</span>,<span class="dv">5</span>:<span class="dv">8</span>),<span class="dt">n.ped=</span><span class="dv">1000</span>, <span class="dt">N.brk=</span><span class="dv">3</span>, <span class="dt">target.snp=</span><span class="ot">NA</span>,<span class="dt">P0=</span><span class="fl">0.001</span>,<span class="dt">is.OR=</span><span class="ot">FALSE</span>,<span class="dt">risk.exposure=</span> <span class="dv">1</span>,<span class="dt">risk.pathway.unexposed=</span><span class="kw">c</span>(<span class="fl">0.5</span>, <span class="dv">1</span>), <span class="dt">risk.pathway.exposed=</span><span class="kw">c</span>(<span class="fl">0.5</span>, <span class="dv">1</span>), <span class="dt">is.case=</span><span class="ot">TRUE</span>, <span class="dt">e.fr=</span><span class="ot">NA</span>, <span class="dt">pop1.frac=</span><span class="ot">NA</span>, <span class="dt">P0.ratio=</span><span class="dv">1</span>,rcmb.rate, <span class="dt">no_cores=</span><span class="dv">5</span>, <span class="dt">qtl=</span>T)</code></pre></div>
<pre><code>## [1]  72  95 139 145 247 276 279 339
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to qtl1.fam 
## Writing extended MAP file to qtl1.bim 
## Writing BED file to qtl1.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to qtl8.fam 
## Writing extended MAP file to qtl8.bim 
## Writing BED file to qtl8.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to qtl17.fam 
## Writing extended MAP file to qtl17.bim 
## Writing BED file to qtl17.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to qtl20.fam 
## Writing extended MAP file to qtl20.bim 
## Writing BED file to qtl20.bed (SNP-major mode)</code></pre>
<p>The following call simulates a scenario that involves gene-environment interaction. The relative risk for the exposure main effect is 1.2. The relative risks for carrying the first and second pathway SNPs are 1.5 and 2 respectively for the exposed individuals and are 1 and 1 for the unexposed individuals.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">TriadSim</span>(input.plink.file, <span class="st">'gxe'</span>, <span class="dt">fr.desire=</span><span class="fl">0.3</span>,<span class="dt">pathways=</span><span class="kw">list</span>(<span class="dv">1</span>:<span class="dv">4</span>,<span class="dv">5</span>:<span class="dv">8</span>),<span class="dt">n.ped=</span><span class="dv">1000</span>, <span class="dt">N.brk=</span><span class="dv">3</span>, <span class="dt">target.snp=</span><span class="ot">NA</span>,<span class="dt">P0=</span><span class="fl">0.001</span>,<span class="dt">is.OR=</span><span class="ot">FALSE</span>,<span class="dt">risk.exposure=</span> <span class="fl">1.2</span>,<span class="dt">risk.pathway.unexposed=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="dt">risk.pathway.exposed=</span><span class="kw">c</span>(<span class="fl">1.5</span>, <span class="dv">2</span>), <span class="dt">is.case=</span><span class="ot">TRUE</span>, <span class="dt">e.fr=</span><span class="fl">0.3</span>, <span class="dt">pop1.frac=</span><span class="ot">NA</span>, <span class="dt">P0.ratio=</span><span class="dv">1</span>,rcmb.rate, <span class="dt">no_cores=</span><span class="dv">5</span>, <span class="dt">qtl=</span><span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## [1]  72  95 139 145 247 276 279 339
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to gxe1.fam 
## Writing extended MAP file to gxe1.bim 
## Writing BED file to gxe1.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to gxe8.fam 
## Writing extended MAP file to gxe8.bim 
## Writing BED file to gxe8.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to gxe17.fam 
## Writing extended MAP file to gxe17.bim 
## Writing BED file to gxe17.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to gxe20.fam 
## Writing extended MAP file to gxe20.bim 
## Writing BED file to gxe20.bed (SNP-major mode)</code></pre>
<p>The following call simulates a stratified scenario that involves gene-environment interaction. The risk parameters are the same as the scenario above. The “input.plink.file”&quot; is a list of two character vectors. Each vector contains three character strings giving the directory and basename of the PLINK files in one subpopulation.The subpopulations are equally sized (pop1.frac=0.5). The baseline disease prevalence (disease prevalence in the unexposed who carries 0 copy of the risk pathway) is 0.001 in the first subpopulation while that in the second subpopulation is 0.003 (0.001*3). The exposure prevalence in the two subpopulations are 0.1 and 0.3 respectively.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(TriadSim)
 m.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_mom'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 f.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_dad'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 k.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_kid'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 m.file2 &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop2_4chr_mom'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 f.file2 &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop2_4chr_dad'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 k.file2 &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop2_4chr_kid'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 input.plink.file2 &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw">c</span>(m.file, f.file, k.file),<span class="kw">c</span>(m.file2, f.file2, k.file2))

 <span class="kw">TriadSim</span>(input.plink.file2, <span class="dt">out.put.file=</span><span class="st">'stratified'</span>, <span class="dt">fr.desire=</span><span class="fl">0.3</span>,<span class="dt">pathways=</span><span class="kw">list</span>(<span class="dv">1</span>:<span class="dv">4</span>,<span class="dv">5</span>:<span class="dv">8</span>),<span class="dt">n.ped=</span><span class="dv">1000</span>, <span class="dt">N.brk=</span><span class="dv">3</span>, <span class="dt">target.snp=</span><span class="ot">NA</span>,<span class="dt">P0=</span><span class="fl">0.001</span>,<span class="dt">is.OR=</span><span class="ot">FALSE</span>,<span class="dt">risk.exposure=</span> <span class="fl">1.2</span>,<span class="dt">risk.pathway.unexposed=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>), <span class="dt">risk.pathway.exposed=</span><span class="kw">c</span>(<span class="fl">1.5</span>, <span class="dv">2</span>), <span class="dt">is.case=</span><span class="ot">TRUE</span>, <span class="dt">e.fr=</span><span class="kw">c</span>(<span class="fl">0.1</span>, <span class="fl">0.3</span>), <span class="dt">pop1.frac=</span><span class="fl">0.5</span>, <span class="dt">P0.ratio=</span><span class="dv">3</span>,rcmb.rate, <span class="dt">no_cores=</span><span class="ot">NA</span>)</code></pre></div>
<pre><code>## [1]  72  95 139 145 247 276 279 339
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to stratified1.fam 
## Writing extended MAP file to stratified1.bim 
## Writing BED file to stratified1.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to stratified8.fam 
## Writing extended MAP file to stratified8.bim 
## Writing BED file to stratified8.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to stratified17.fam 
## Writing extended MAP file to stratified17.bim 
## Writing BED file to stratified17.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to stratified20.fam 
## Writing extended MAP file to stratified20.bim 
## Writing BED file to stratified20.bed (SNP-major mode)</code></pre>
<p>To simulate case-control data the function needs to be called twice, calls to simulate cases (is.case=TRUE) and controls (is.case=FALSE) respectively. The script below calls the function to simulate 1000 cases and 1000 controls and writes genotypes of the cases and controls into seperate sets of PLINK files.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## cases
 <span class="kw">TriadSim</span>(input.plink.file, <span class="st">'case'</span>, <span class="dt">fr.desire=</span><span class="fl">0.05</span>,<span class="dt">pathways=</span><span class="kw">list</span>(<span class="dv">1</span>:<span class="dv">4</span>,<span class="dv">5</span>:<span class="dv">8</span>),<span class="dt">n.ped=</span><span class="dv">1000</span>, <span class="dt">N.brk=</span><span class="dv">3</span>, <span class="dt">target.snp=</span><span class="ot">NA</span>,<span class="dt">P0=</span><span class="fl">0.001</span>,<span class="dt">is.OR=</span><span class="ot">TRUE</span>,<span class="dt">risk.exposure=</span> <span class="dv">1</span>,<span class="dt">risk.pathway.unexposed=</span><span class="kw">c</span>(<span class="fl">1.5</span>, <span class="dv">2</span>), <span class="dt">risk.pathway.exposed=</span><span class="kw">c</span>(<span class="fl">1.5</span>, <span class="dv">2</span>), <span class="dt">is.case=</span><span class="ot">TRUE</span>, <span class="dt">e.fr=</span><span class="ot">NA</span>, <span class="dt">pop1.frac=</span><span class="ot">NA</span>, <span class="dt">P0.ratio=</span><span class="dv">1</span>,rcmb.rate, <span class="dt">no_cores=</span><span class="ot">NA</span>)</code></pre></div>
<pre><code>## [1]  21 118 121 140 155 168 218 383
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to case1.fam 
## Writing extended MAP file to case1.bim 
## Writing BED file to case1.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to case8.fam 
## Writing extended MAP file to case8.bim 
## Writing BED file to case8.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to case17.fam 
## Writing extended MAP file to case17.bim 
## Writing BED file to case17.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to case20.fam 
## Writing extended MAP file to case20.bim 
## Writing BED file to case20.bed (SNP-major mode)</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## controls
 <span class="kw">TriadSim</span>(input.plink.file, <span class="st">'ctrl'</span>, <span class="dt">fr.desire=</span><span class="fl">0.05</span>,<span class="dt">pathways=</span><span class="kw">list</span>(<span class="dv">1</span>:<span class="dv">4</span>,<span class="dv">5</span>:<span class="dv">8</span>),<span class="dt">n.ped=</span><span class="dv">1000</span>, <span class="dt">N.brk=</span><span class="dv">3</span>, <span class="dt">target.snp=</span><span class="ot">NA</span>,<span class="dt">P0=</span><span class="fl">0.001</span>,<span class="dt">is.OR=</span><span class="ot">TRUE</span>,<span class="dt">risk.exposure=</span> <span class="dv">1</span>,<span class="dt">risk.pathway.unexposed=</span><span class="kw">c</span>(<span class="fl">1.5</span>, <span class="dv">2</span>), <span class="dt">risk.pathway.exposed=</span><span class="kw">c</span>(<span class="fl">1.5</span>, <span class="dv">2</span>), <span class="dt">is.case=</span><span class="ot">FALSE</span>, <span class="dt">e.fr=</span><span class="ot">NA</span>, <span class="dt">pop1.frac=</span><span class="ot">NA</span>, <span class="dt">P0.ratio=</span><span class="dv">1</span>,rcmb.rate, <span class="dt">no_cores=</span><span class="ot">NA</span>)</code></pre></div>
<pre><code>## [1]  21 118 121 140 155 168 218 383
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to ctrl1.fam 
## Writing extended MAP file to ctrl1.bim 
## Writing BED file to ctrl1.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to ctrl8.fam 
## Writing extended MAP file to ctrl8.bim 
## Writing BED file to ctrl8.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to ctrl17.fam 
## Writing extended MAP file to ctrl17.bim 
## Writing BED file to ctrl17.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to ctrl20.fam 
## Writing extended MAP file to ctrl20.bim 
## Writing BED file to ctrl20.bed (SNP-major mode)</code></pre>
<div id="some-additional-details" class="section level3">
<h3>Some additional details</h3>
<p>The source data may contain genotyping errors that cause non-Mendelian inheritance patterns. For these non-Mendelian families, genotypes of the three individuals in the family will be set to missing at the corresponding SNPs. We assume nonpaternity and adoption have both been ruled out in QC for the source data.</p>
<p>This function requires at least two pathway SNPs, eithe two SNPs in the one pathway or two pathways each involving one SNP. If the users are interested in a single SNP scenario one can trick the function by setting the number of pathway to 2, each with a single SNP in the pathway but only the SNP in the first pathway carries a risk while that in the second pathway does not change risk. See below for an example.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">TriadSim</span>(input.plink.file, <span class="st">'singleSNP'</span>, <span class="dt">fr.desire=</span><span class="fl">0.05</span>,<span class="dt">pathways=</span><span class="kw">list</span>(<span class="dv">1</span>,<span class="dv">2</span>),<span class="dt">n.ped=</span><span class="dv">1000</span>, <span class="dt">N.brk=</span><span class="dv">3</span>, <span class="dt">target.snp=</span><span class="ot">NA</span>,<span class="dt">P0=</span><span class="fl">0.001</span>,<span class="dt">is.OR=</span><span class="ot">FALSE</span>,<span class="dt">risk.exposure=</span> <span class="dv">1</span>,<span class="dt">risk.pathway.unexposed=</span><span class="kw">c</span>(<span class="fl">1.5</span>, <span class="dv">1</span>), <span class="dt">risk.pathway.exposed=</span><span class="kw">c</span>(<span class="fl">1.5</span>, <span class="dv">1</span>), <span class="dt">is.case=</span><span class="ot">TRUE</span>, <span class="dt">e.fr=</span><span class="ot">NA</span>, <span class="dt">pop1.frac=</span><span class="ot">NA</span>, <span class="dt">P0.ratio=</span><span class="dv">1</span>,rcmb.rate, <span class="dt">no_cores=</span><span class="ot">NA</span>)</code></pre></div>
<pre><code>## [1] 118 140
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to singleSNP1.fam 
## Writing extended MAP file to singleSNP1.bim 
## Writing BED file to singleSNP1.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to singleSNP8.fam 
## Writing extended MAP file to singleSNP8.bim 
## Writing BED file to singleSNP8.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to singleSNP17.fam 
## Writing extended MAP file to singleSNP17.bim 
## Writing BED file to singleSNP17.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to singleSNP20.fam 
## Writing extended MAP file to singleSNP20.bim 
## Writing BED file to singleSNP20.bed (SNP-major mode)</code></pre>
</div>
</div>
<div id="facility-functions" class="section level2">
<h2>Facility functions</h2>
<p>The following set of functions is provided in case users want to have more control over the simulation parameters. They are called by the main function to generate simulations. Users do not need to call them directly.</p>
<div id="pick_target.snp" class="section level3">
<h3><code>pick_target.snp</code></h3>
<p>Users can manually pick the target SNPs in the pathway or use the facility function <code>pick_target.snp</code> to pick the set of target SNPs in the pathway(s) based on a desired allele frequency. The example below uses the example files that come with the package to select 8 SNPs with allele frequencies close to 0.05. The function returns the selected target SNPs by giving the row numbers (i.e., the order) of the corresponding SNPs among all the SNPs in the associated “bim” file. For example a return of “1084 2044 3285 4016 5117 6067 7077 8187” means the SNPs at rows 1084 2044 3285 4016 5117 6067 7077 8187 are selected to be the target SNPs in the pathway.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> m.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_mom'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 f.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_dad'</span>,<span class="dt">sep=</span><span class="st">''</span>)
 picked.target &lt;-<span class="st"> </span><span class="kw">pick_target.snp</span>(<span class="kw">c</span>(m.file,f.file),<span class="fl">0.05</span>, <span class="dv">8</span>)</code></pre></div>
<pre><code>## [1]  21 118 121 140 155 168 218 383</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"> <span class="kw">cat</span>(<span class="st">'Target SNPs picked:'</span>,picked.target[[<span class="dv">2</span>]],<span class="st">'</span><span class="ch">\n</span><span class="st">'</span>)</code></pre></div>
<pre><code>## Target SNPs picked: 21 118 121 140 155 168 218 383</code></pre>
</div>
<div id="get.target.geno" class="section level3">
<h3><code>get.target.geno</code></h3>
<p>The function <code>get.target.geno</code> retrieves genotypes of the target SNPs and returns the genotypes of the triads in a list of three elements: the observed genotypes of the mothers from family 1 to family n repeated twice, genotypes of the fathers from family 1 to family n repeated twice and genotypes of children from family 1 to n followed by (stacking on top of) genotypes of the complements in the same family order.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">target.snp &lt;-<span class="st"> </span>picked.target[[<span class="dv">2</span>]]
m.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_mom'</span>,<span class="dt">sep=</span><span class="st">''</span>)
f.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_dad'</span>,<span class="dt">sep=</span><span class="st">''</span>)
k.file &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="kw">system.file</span>(<span class="dt">package =</span> <span class="st">&quot;TriadSim&quot;</span>),<span class="st">'/extdata/pop1_4chr_kid'</span>,<span class="dt">sep=</span><span class="st">''</span>)
<span class="co"># the preloaded data frame snp.all2 contains the data frame read from the corresponding .bim file.</span>
target.geno &lt;-<span class="st"> </span><span class="kw">get.target.geno</span>(<span class="kw">c</span>(m.file,f.file,k.file), target.snp,snp.all2)</code></pre></div>
<p>The output <em>target.geno</em> is a list of three elements, each being a matrix of genotypes</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">length</span>(target.geno)</code></pre></div>
<pre><code>## [1] 3</code></pre>
<p>For this example the genotypes form a 2000 x 8 numerical matrix (2x1000 families and 8 SNPs)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mom.target &lt;-<span class="st"> </span>target.geno[[<span class="dv">1</span>]]
dad.target &lt;-<span class="st"> </span>target.geno[[<span class="dv">2</span>]]
kid.target &lt;-<span class="st"> </span>target.geno[[<span class="dv">3</span>]]
<span class="kw">str</span>(mom.target)</code></pre></div>
<pre><code>##  num [1:2000, 1:8] 1 2 2 2 2 2 2 2 2 0 ...</code></pre>
<p>To increase diversity, TriadSim introduces break points at each chromosome, selecting them independently for each triad being simulated. The break points can be picked manually or using the function <code>get.brks</code>. The function tends to pick the break points at recombination hotspots if such data are passed in as an input parameter <em>rcmb.rate</em>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">found.brks &lt;-<span class="st"> </span><span class="kw">get.brks</span>(<span class="dt">N.brk=</span><span class="dv">3</span>,<span class="dt">n.ped=</span><span class="dv">1000</span>, snp.all2, target.snp,<span class="dt">rcmb.rate=</span>rcmb.rate)
breaks &lt;-<span class="st"> </span>found.brks[[<span class="dv">1</span>]]
family.position &lt;-<span class="st"> </span>found.brks[[<span class="dv">2</span>]] </code></pre></div>
<p>This function returns a list of two items. The first is a 1000 x 17 matrix of integers showing where the chromosomal breaks are to take place (in terms of the order of the SNPs in the PLINK files) for each individual in the simulated trios. Each chromosome has 3 breaks, adding to that is the number of breaks between chromosomes, i.e., 3, and the first and the last SNPs, and this is where the 17 comes from. Here 1000 denotes the number of triads in the simulated data as defined by the <code>n.ped</code> input parameter.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(breaks)</code></pre></div>
<pre><code>## [1] 1000   19</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(breaks)</code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12] [,13]
## [1,]    0   40  118  130  154  162  173  183  243   259   297   303   326
## [2,]    0   69  120  123  152  159  173  181  215   264   297   312   320
## [3,]    0   35  118  129  143  162  173  202  233   257   297   310   320
## [4,]    0   86  120  122  152  161  173  191  232   264   297   309   326
## [5,]    0   58  120  135  147  162  173  204  235   266   297   309   326
## [6,]    0   58  120  123  152  165  173  209  248   264   297   313   320
##      [,14] [,15] [,16] [,17] [,18] [,19]
## [1,]   352   355   363   377   408   412
## [2,]   339   355   369   381   398   412
## [3,]   348   355   367   381   408   412
## [4,]   337   355   372   377   407   412
## [5,]   337   355   371   384   405   412
## [6,]   337   355   372   379   398   412</code></pre>
<p>The second one is a 1000 x 8 matrix showing the chromosomal segments out of which each target SNP is selected for each simulated trio.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(family.position)</code></pre></div>
<pre><code>## [1] 1000    8</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(family.position)</code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
## [1,]    1    2    3    4    5    6    8   17
## [2,]    1    2    3    4    5    6    9   17
## [3,]    1    2    3    4    5    6    8   17
## [4,]    1    2    3    4    5    6    8   17
## [5,]    1    2    3    4    5    6    8   16
## [6,]    1    2    3    4    5    6    8   17</code></pre>
</div>
<div id="fit.risk.model.par" class="section level3">
<h3><code>fit.risk.model.par</code></h3>
<p><code>fit.risk.model.par</code> is a function that resamples families based on the specified risk model. It can simulate a homogenous scenario or a stratified scenario with two subpopulations. The risk model can involve exposure main effects as well as gene by exposure interactions. This function is parallelized to shorten the running time. An example call for simulating a binary phenotype is given below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">betas &lt;-<span class="st"> </span><span class="kw">c</span>(-<span class="fl">6.4</span>, <span class="fl">3.2</span>, <span class="fl">5.8</span>)
pwy &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="dv">1</span>:<span class="dv">4</span>,<span class="dv">5</span>:<span class="dv">8</span>)
## scenarios of genetic main effects only for a binary phenotype
fitted.model1 &lt;-<span class="st"> </span><span class="kw">fit.risk.model.par</span>(<span class="dt">n.ped=</span><span class="dv">1000</span>,<span class="dt">brks=</span>breaks,target.snp,<span class="dt">fam.pos=</span>family.position, 
<span class="dt">mom.tar=</span>mom.target,<span class="dt">dad.tar=</span>dad.target, <span class="dt">kid.tar=</span>kid.target, <span class="dt">pathways=</span>pwy, 
betas, <span class="dt">e.fr=</span><span class="ot">NA</span>, betas,<span class="dt">pop1.frac=</span> <span class="ot">NA</span>,<span class="dt">rate.beta=</span><span class="ot">NA</span>,<span class="dt">qtl=</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>This function returns a list of five items. For a scenario involving a binary trait in a homogeneous population and no gene-environment interaction only the first two items contain the data needed. The first is a 1000 x 16 matrix of integers showing which source families are picked for each chromosomal segments in each of the 1000 simulated trios.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sel.fam &lt;-<span class="st"> </span>fitted.model1[[<span class="dv">1</span>]]
<span class="kw">colnames</span>(sel.fam) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">'seg'</span>,<span class="dv">1</span>:<span class="kw">ncol</span>(sel.fam),<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)
<span class="kw">rownames</span>(sel.fam) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">'fam'</span>,<span class="dv">1</span>:<span class="kw">nrow</span>(sel.fam),<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)
<span class="kw">dim</span>(sel.fam)</code></pre></div>
<pre><code>## [1] 1000   18</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(sel.fam)</code></pre></div>
<pre><code>##       seg_1 seg_2 seg_3 seg_4 seg_5 seg_6 seg_7 seg_8 seg_9 seg_10 seg_11
## fam_1  1972  1567  1381   956  1846  1761   648  1220   540    444   1204
## fam_2   627   364   621  1732  1330    28  1163   748   583     75   1779
## fam_3  1558    15  1954   903   415   440  1185  1648  1931   1549    349
## fam_4  1451   900   122   516   609   863  1474   344   270    518    448
## fam_5  1373  1572  1129     9   504   659  1373  1854  1841   1658    483
## fam_6  1182    48  1576  1431   102  1809   394   995   858     52   1250
##       seg_12 seg_13 seg_14 seg_15 seg_16 seg_17 seg_18
## fam_1    183   1436    390     76    805   1352   1984
## fam_2    349    699    535   1515   1987    963    903
## fam_3    945    858   1094    441    748   1049   1257
## fam_4    556   1954   1630   1996   1181    746    531
## fam_5    899   1653     20   1814   1943   1779    649
## fam_6    338   1847   1823    889    254   1572    514</code></pre>
<p>The second one is a 1000 x 8 matrix showing the genotypes at the 8 target SNPs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sim.pathway.geno &lt;-<span class="st">  </span>fitted.model1[[<span class="dv">2</span>]]
<span class="kw">colnames</span>(sim.pathway.geno) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">'target.snp'</span>,<span class="dv">1</span>:<span class="kw">ncol</span>(sim.pathway.geno),<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)
<span class="kw">rownames</span>(sim.pathway.geno) &lt;-<span class="st"> </span><span class="kw">paste</span>(<span class="st">'fam'</span>,<span class="dv">1</span>:<span class="kw">nrow</span>(sim.pathway.geno),<span class="dt">sep=</span><span class="st">&quot;_&quot;</span>)
<span class="kw">dim</span>(sim.pathway.geno)</code></pre></div>
<pre><code>## [1] 3000    8</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(sim.pathway.geno)</code></pre></div>
<pre><code>##       target.snp_1 target.snp_2 target.snp_3 target.snp_4 target.snp_5
## fam_1            2            2            2            2            2
## fam_2            1            1            2            2            2
## fam_3            2            2            2            0            2
## fam_4            2            2            2            2            2
## fam_5            2            2            2            2            2
## fam_6            0            2            2            2            2
##       target.snp_6 target.snp_7 target.snp_8
## fam_1            2            2            2
## fam_2            2            0            2
## fam_3            2            2            2
## fam_4            2            2            2
## fam_5            2            2            2
## fam_6            2            0            2</code></pre>
<p>The following is an example call for a scenario involving gene-environment interaction for a binary phenotype.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## a scenario of gene-environment interaction  for a binary phenotype
betas.e &lt;-<span class="st"> </span><span class="kw">c</span>(-<span class="fl">6.4</span>, <span class="fl">3.9</span>, <span class="fl">6.5</span>)

fitted.model2 &lt;-<span class="st"> </span><span class="kw">fit.risk.model.par</span>(<span class="dt">n.ped=</span><span class="dv">1000</span>,<span class="dt">brks=</span>breaks,target.snp,<span class="dt">fam.pos=</span>family.position, 
<span class="dt">mom.tar=</span>mom.target,<span class="dt">dad.tar=</span>dad.target, <span class="dt">kid.tar=</span>kid.target, <span class="dt">pathways=</span>pwy, 
betas, <span class="dt">e.fr=</span> <span class="fl">0.2</span>, betas.e,<span class="dt">pop1.frac=</span> <span class="ot">NA</span>,<span class="dt">rate.beta=</span><span class="ot">NA</span>, <span class="dt">qtl=</span> <span class="ot">FALSE</span>)</code></pre></div>
<p>For this scenario the third returned item contains exposure data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">exposure &lt;-<span class="st">  </span>fitted.model2[[<span class="dv">3</span>]]
<span class="kw">table</span>(exposure)</code></pre></div>
<pre><code>## exposure
##   0   1 
## 795 205</code></pre>
<p>The following is an example call for a quantitative trait scenario.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## scenarios of a quantitative trait phenotype
fitted.model3 &lt;-<span class="st"> </span><span class="kw">fit.risk.model.par</span>(<span class="dt">n.ped=</span><span class="dv">1000</span>,<span class="dt">brks=</span>breaks,target.snp,<span class="dt">fam.pos=</span>family.position, 
<span class="dt">mom.tar=</span>mom.target,<span class="dt">dad.tar=</span>dad.target, <span class="dt">kid.tar=</span>kid.target, <span class="dt">pathways=</span>pwy, 
betas, <span class="dt">e.fr=</span><span class="ot">NA</span>, betas,<span class="dt">pop1.frac=</span> <span class="ot">NA</span>,<span class="dt">rate.beta=</span><span class="ot">NA</span>,<span class="dt">qtl=</span><span class="ot">TRUE</span>)</code></pre></div>
<p>For this scenario the fifth returned item contains data for the quantitative phenotype.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">qt.pheno &lt;-<span class="st">  </span>fitted.model3[[<span class="dv">5</span>]]
<span class="kw">hist</span>(qt.pheno,<span class="dt">main=</span><span class="st">'Histogram of Simulated Quantitative Trait'</span>,<span class="dt">xlab=</span><span class="st">'QT'</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAkAAAAJACAMAAABSRCkEAAAAbFBMVEUAAAAAADoAAGYAOpAAZrY6AAA6ADo6AGY6OgA6Ojo6kNtmAABmADpmZmZmtv+QOgCQOjqQkGaQtpCQ29uQ2/+2ZgC2Zjq2Zma2/7a2/9u2///bkDrb25Db/7bb////tmb/25D//7b//9v/////yF58AAAACXBIWXMAAA7DAAAOwwHHb6hkAAATeklEQVR4nO2dfWOyuhmH7VO7l7rVvdRt57DVqt//Ow4IwSiCgTsh6Z3r+uMcnyq/xHAZkoiwuQAI2KSuAPxsEAhEIBCIQCAQgUAgAoFABAKBCAQCEQgEIhAIRCAQiEAgEIFAIAKBQAQCgQgEAhEIBCIQCEQgEIhAIBCBQCACgUAEAoEIBAIRCAQiEAhEIBCIQCAQgUAgAoFABAKBCAQCEWkEOu83r1/m/9v+Hw7//j10iYfNpi7KUNWP+/Lvix6p6dPKHTa/rn8/NiVsPvyrV2fakgbxznOjHDY92wdPj2bLyVGg0+5X6Dfa7tF3W/bG7t8FAo1VzhHotOtK8H0bbaYpaRDvPDeBl0AR2jULgQYcvFvem2rz8tk/nLN7B7turHLXv/eGehvkZA7i/RrjiUBzomaSgUDdP/oDy8E2/fdb3200f3z5bFugduEfb40P7dONF/Unfntsu5RqcxXl4gS0kd0zh/ZBs5c/TNHDzU1L139/t5Xry+or5xwTr3VrqbqjV7PN+02Yk9MqfXAzf2tKsvF35XXPtRvWT9m3dOtK/2kcttDZzQ5LNgJV/WfWvtHuL80LzYf65U/XP79+mXFG0z79IePuU38NuBWofUW7H6xAd5sPBLqW1VWu26Z5Wd/hdAXX/+5Kql/Uln0V6JrTd4R9pivQfXntc0fjZeW85RsdHIHuW0ijQNcet+8G2gbuP7P1HjZemYbfth86I1Dn1LbToGnN90vfiXR91k3A3SHM/qMX6Hbze4Gcsrrn2v/2UvR1azDvpOVgFHDD+pzKvhmbaVS1D2/Lc5rIeVnXqzmN2gt010LORqHJRaD6fd6NA7q3ezCdUrcTTJv305uqDTCfdPPfroO/xlzsce/zruyPy1Wgu82Hh7C+rJtdV93XrcGpQXUnkJtj3kb94u1QoEF5tpM2B7f3m/LdRu0FumuhAgTq/tI2stNmZifYxzcu2Ixt95k3/73uPjfAFag/kNx+st3NBwJdyzLPtYMLc3iwu+3gI5CTU/XDtwcC3ZdnnquTP+rK35Z/06jba7E3LaRRoPsxkB2JfFw/5G1zNEd++9gVqG0d88xjgdyAW4EunUPbaYGunb9TlnnuuOl34E3dnHKrNr3voWxYnzMl0KC8XuNt/e/txS3/plFvBHJSShDocp2HefVAZldPCDTSA/UvaMI8BXLK6nugj7u30u8dM4iuX1yP+a+7rY1wc6YEGpTXT1R//WfXHfmGq5T3AjkpJQh0/ls/gX42Bmqb52jHEO8jAo2MgZqe7uPShY0IZF7djujP1/mPI5czUB6MgbppvOlRb8PcnCmBBuV17XXavfzdljZc7rkXyEkpQCAzZbKDxaYBxmdh9vNlji1jPdDULOzBBPBWoO548+5UzpTVJR06R8wU/2YW5q4qdOtbNmyY4/RGznsfvO7azW3sAbMv323U+x7IprjtGphcBLL7tfu0OutA28v9OtD1CD81iL64AQ9mYROD6H6M6o6BuihTuU6SbtmxzzM4Bg3CHg6i28zf+hWn5uFteb917XW0ox63fKdRh2MgdxB93GhaBxqMgdqGNu+vW6Ft/9Id69tP+GAW9vJZmb35UCAn4MEs7N3W48HmzYav/7udhZmybOXavdN9/huDPh58mfrXq0EmzM1xBWoz/7m36/DtoO+mvO45p89xy7826nAWZlLOTnawnWj4UadzHJ599Zkbhx9W3wX8DIEO9qM8/k0hpOFnCGTXPSIMAkHGzxCoGzTqPyD8PH6IQJArCAQiEAhEIBCIQCAQgUAgAoFABAKBCAQCEQgEIhAIRCAQiEAgEIFAIAKBQAQCgQgEAhEIBCIQCEQgEIhAIBCBQCACgUAEAoEIBAIRCAQiEAhEIBCIQCAQgUAgAoFABAKBCAQCEQgEIhAIRCAQiEAgEIFAIAKBQAQCgQgEAhEIBCIQKGM2E6SumyWbisCQiZ2TzX7LpiIwBIFABAKBCAQCEQgEIhAIRCAQiEAgEIFAIAKBQAQCgQgEAhH6BDrvzVfBv36PVB1wUSdQtXk3D472AcREm0Dnfa9N9foVoTJwizaBTrsP+/DIQWwFtAlED7Qy2gSqx0BdF8QYaBXUCVQfxMwsjP5nFfQJBKuCQCBCm0CnXTPyObKQuBYqBWrnX86EHuKhUaBOHabxa6BRoO+3VqC7hcT8fjCpAo0CPemBsnlfKtAnUNPHbC92OC2Mg2doE+jSOvTyObEQnc37UoFCgVaOKxwEAhH6BKrqMZAZRI8sJWbzvlSgTqCqHv+cds0oGoHWQJtA5nyg876ewiPQGmgTyH6BcXj9QqA10CZQf0biYYtAa6BNoP7AddqNfR+fzftSgTqB+nNaz3sEWgF9Aq0dVzgIBCIQCEQgEIhAIBCBQCACgUAEAoEIBAIRCAQiEAhEIBCIQCAQgUAgAoFABAKBCAQCEQgEIhAIRCAQiEAgEIFAIAKBQAQCgQgEAhEIBCIQCEQgEIhAIBCBQCACgUAEAoEIBAIRCAQiEAhEIBCIQCAQgUAgAoFABAKBCAQCEQgEIhAIRCAQiNAn0Hm/aRm51UpG70sF6gSq7O2+R+/7nc37UoE2gfobztUqvX6J4+AZ2gSyt7ysOXK/sBXQJhA90MpoE6i/YSFjoHVQJ1Bzs9SWkf4no/elAn0CrR1XOAgEIvQJxELiqqgTiIXEddEmENP4ldEm0PhC4qYnWMVAn0D0QCujTSAWEldGnUAsJK6LPoHWjiscBAIR+gSq6sNXOwyqOJ1jBdQJVL181sOg7QWB1kGbQGYaf97XQ2gEWgNtAtmFxMPrFwKtgTaB+oXEwxaB1kCbQP2B67Qb+z4+m/elAnUC9UvR5z0CrYA+gdaOKxwEAhEIBCIQCEQgEIhAIBCBQCACgUAEAoEIBAIRCAQiEAhEIBCImBJonPXq11Yk67jCWdQDIRBYEAhEIBCIQCAQgUAgAoFABAKBCAQCEQgEIhAIRCAQiEAg8GDR96IIBJZFLiAQWBAoXlwRIFC8uCJAoHhxRYBA8eKKAIHixRUBAsWLKwIEihdXBAgUL64IECheXBEgULy4IkCgeHFFgEDx4opAiUCn3WYbMA68USKQuRnYyK1Ql8SBJ2oEugRwCIHmo0kgo9DIfTCWxMFzFAl0bO9o2d5R7jHnvTnVclQxBJqPFoGaW3obc45jflT2+MZ94wOiRKDT7uXzyWv7G87VKo10Ugg0HyUCeWBveXkZ76QQaD5qBDrUvcrosamBHigKWgQ6tE6Yu3qPYG9YyBgoJEoEssen0RG0eZGZhY1O0xBoPkoEssensfsxz4wDb5QI1B2fvt9YiV4ZLQLV7tQHp6m5/GnXyHVkITEsagR6SitQO/9yJvTL48BQlkCdOkzjw6FFoKffc7UCfb+1At3N1VLdqEEFWgQ6PJ1+0QNFQYlAZoQ8iVkF2k69GIHmo0agkXHxDbVD9Txt/AsPBJqPEoHO+wCnRCPQApQIVHcrPl2Qdxz4okQg+zWX5IRWBFqCEoGyjCuCcgTqO6nxfgqB5qNGoNqP16/D1GT+vH92fEOg+WgR6PjyWb1+TS8HPZ2qIdB8lAjUnA/UrC9Pnw/0bKqGQPNRIlCzkNgINHlGon8ceKNEINsDHUZPV50VB94oEagbA1Wy5UQEmo8Wgcws/emvC73jwBM1AmUYVwQIFC+uCJQIxHdhqVAiUIdsFo9AC9Al0OUgOi0IgeajTCAWEtdGmUD8tHltdAk0eXWO+XHwHCUCPb3wxrw48EaJQFnGFQECxYsrAiUCeZyvOicOvFEiUHeuGN/Gr44Sgc57Y87Yr95nxoE3SgQ6/cWcyMFC4tooEcj2QJyRuDZKBGrOSKz/W8nOKEOg+WgRyMzDZOuICLQANQJlGFcECBQvrgjUCPT8p82z4sATLQL5/LR5Rhz4okQgv582e8eBN0oE4qfNqVAiED9tToUSgfhpcyq0CMRPmxOhRqAM44pAiUDODVFDxIE3SgTyu1K9dxx4o0Qg6Y+a7+PAFyUCcXGFVCgRKMu4IkCgeHFFoEGgMCPoCwItQY1AISbyCDQfBBrEwRwQaBAHc0CgQRzMAYEGcTCHkgR6em95BJqPDoG8Ls1R2bs1j962GYHmo0EgP5wuauwSDAg0n3IEcpYbx755RaD5lCMQPVAUyhHoevUpxkABKUig55dyRaD5lCTQ2nFFgEDx4oqgJIFYSIxAQQKxkBiDcgRiGh+FcgQaX0i8fhESrGLlUI5A9EBRKEcgFhKjUJBALCTGoCSB1o4rAgSKF1cECBQvrggQKF5cEZQjkMeJrwg0n3IEupz3zy7dgUDzKUig2qAnNwVHoPmUJFB3X8xgcXApTKCV44oAgeLFFQECxYsrAgSKF1cECBQvrggQKF5cESBQvLgiQKB4cUWAQPHiigCB4sUVAQLFiysCBIoXVwQIFC+uCBAoXlwRIFC8uCJAoHhxRYBA8eKKAIHixRUBAsWLKwIEihdXBAgUL64IECheXBEgULy4IkCgeHFFgEDx4ooAgeLFFQECxYsrAgSKF1cECBQvrggQKF5cESBQvLgiQKB4cUWAQPHiigCB4sUVAQLFiysCBIoXVwQIFC+uCBAoXlwRIFC8uCJAoHhxRYBA8eKKAIHixSliM874NhNxSzaKAQKtxJI9jkDQg0Ap4hSBQCnifhoTAx0EShH30wi8xxGoNBAos7ifBgI94byfvOUuAi15qiCBqs27eXC0D0Rx+kCgSc77Xpvq9UscpxAEmuS06++4e+S+8Y9AoEnogZ6RhUBL1qKWM3MM1HVBjIEek4VASzZazrzM086YPNL/INCSp0oSaO24PFl0jECgyZSYR9nsWG2PaxPotGtGPsfiFxIRaGFmK1A7/3Im9MvjfiwItDCzEahTR/80PvCZGQh0MQJ9v7UC6V9IXG/nlSVQOT3QkqcQaBKzCrS92OG0MC5vEMiPmZm1Qy+f4wvRCBR0I4UCrRyXEATyA4FGQCA/EGgEBPIDgUZAID8QaAQE8gOBRkAgPxBoBATyA4FGQCA/EGgEBPKjcIFWO7kQgZLERSeHnZdDHZaDQLOfQaCImQgUciMEyp4cdl4OdVhOCQLlcHYqAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJC0MOLiBQkrgw5OACAiWJm1V01hcbQ6AkcWGKzsEFBEoSF6boHFxAoCRxYYrOwQUEShIXpugcXECgJHFhis7BBQRKEhem6BxcQKAkccP8n3q5QwRKEjcnP4f9gECZxc3Jz2E/IFBmcXPyc9gPCJRHnMJvJfIQaEm7PiFPgZbkK9wohzo8A4Ey3iiHOjwDgTLeKIc6PGPelue9OWL++j14RfxSsmhSBFq6ZbV5Nw+O9sHTuMArf1k0KQIt3PK877WpXr/84jQ2KQIt3PK0+7APj7cHsfFeZaoHgnxYRSCPHghKY+YYqOuCRsdAUBrz+q7TznR49D/QEePrESgIBAIRCAQiEAhEIBCIQCAQgUAgAoFABAKBCAQCEQgEImILlPYkBbglxg6OkLlmvg851CGLSiDQMnKoQxaVQKBl5FCHLCqBQMvIoQ5ZVAKBlpFDHbKoBAItI4c6ZFEJBFpGDnXIohIItIwc6pBFJRBoGTnUIYtK/ESBQDkIBCIQCEQgEIhAIBCBQCACgUAEAoEIBAIRCAQiEAhEIBCIQCAQEVWg7z+213I9bjYvnzHLmcZcl2+brgLJW6AhVivEFOi0ay8GfKzb7piw/b7/kHjfJW+BhlitEFGgo7klgrk48CFdD3AcvTHDSiRvgYZYrRBPoOPmva3091tzbeAq3V6s0u65DFqgIVYrRB0DGYHavjNhN3D4c33wT3ld6+Qt0BCrFeILZA7+6YYAp11zVetDQoNSt0BDtFbQL5BTlVRlZ9ECbU3Ct0IBh7AWMw5JVHYWLdDWJHwrxBCoskfb1IPo6nrYTzmXz2IQbWoSvhXi90CpJ7Fm96X8/KdugYZorRBfoOTLaO2eSzmITt4CDbFaYQWBmiNJ0tY71EeydCOghtQt0BCpFfgyFUQgEIhAIBCBQCACgUAEAoEIBAIRCAQiEAhEIBCIQCAQgUAgAoFABAKBCAQCEQgEIhAIRCAQiEAgEIFAIAKBQAQCgQgEAhEIBCIQCEQgEIhAIBCBQP4cN+Yn7uaKuRtzDdHSQSBfzntztSNzgYv01w7OBATypbtAVHeRLwTqQCBPTrvu2jrmQlEI1IFAnvTGmIseIVAHAnnSX2HMPECgDgTyBIEeg0CecAh7DAJ5ch1EN5d8RyALAvnSTOPP+9f/moUgBOpAIG8OrUGbtgNCIAsC+XN0vsBAoA4Ems2/UMcBgUAEAoEIBAIRCAQiEAhEIBCIQCAQgUAgAoFABAKBCAQCEQgEIhAIRCAQiEAgEIFAIAKBQAQCgQgEAhEIBCIQCEQgEIhAIBDxf/ICwmbkBjXVAAAAAElFTkSuQmCC" alt /><!-- --></p>
</div>
<div id="glue.chr.segment.par" class="section level3">
<h3><code>glue.chr.segment.par</code></h3>
<p><code>glue.chr.segment.par</code> is a function that splices the triad chromosomal segments into “complete” trios. The spliced trio sets are written into separate plink files chromosome by chromosome. It is parallelized and if no “no_cores”&quot; value is given half of the total number of CPUs available will be used in the parallelization. Be aware that running the following code will write PLINK files to your current working directory.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">glue.chr.segment.par</span>(<span class="kw">c</span>(m.file,f.file,k.file),<span class="st">'trio'</span>, breaks,sel.fam,snp.all2,sim.pathway.geno,target.snp,<span class="dt">pop.vec=</span><span class="ot">NA</span>,<span class="dt">no_cores=</span><span class="ot">NA</span>,<span class="dt">flip=</span><span class="ot">TRUE</span>) </code></pre></div>
<pre><code>## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to trio1.fam 
## Writing extended MAP file to trio1.bim 
## Writing BED file to trio1.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to trio8.fam 
## Writing extended MAP file to trio8.bim 
## Writing BED file to trio8.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to trio17.fam 
## Writing extended MAP file to trio17.bim 
## Writing BED file to trio17.bed (SNP-major mode)
## coercing object of mode  numeric  to SnpMatrix
## Writing FAM file to trio20.fam 
## Writing extended MAP file to trio20.bim 
## Writing BED file to trio20.bed (SNP-major mode)</code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
